lib/widgets/theme.dart

```diff
@@ -0,0 +1,80 @@
+import 'package:flutter/material.dart';
+
+import 'text.dart';
+
+ThemeData outerThemeData(BuildContext context) {
+  return ThemeData(
+    typography: zulipTypography(context),
+    appBarTheme: const AppBarTheme(
+      // Set these two fields to prevent a color change in [AppBar]s when
+      // there is something scrolled under it. If an app bar hasn't been
+      // given a backgroundColor directly or by theme, it uses
+      // ColorScheme.surfaceContainer for the scrolled-under state and
+      // ColorScheme.surface otherwise, and those are different colors.
+      scrolledUnderElevation: 0,
+      backgroundColor: Color(0xfff5f5f5),
+
+      shape: Border(bottom: BorderSide(color: Color(0xffcccccc))),
+    ),
+    // This applies Material 3's color system to produce a palette of
+    // appropriately matching and contrasting colors for use in a UI.
+    // The Zulip brand color is a starting point, but doesn't end up as
+    // one that's directly used.  (After all, we didn't design it for that
+    // purpose; we designed a logo.)  See docs:
+    //   https://api.flutter.dev/flutter/material/ColorScheme/ColorScheme.fromSeed.html
+    // Or try this tool to see the whole palette:
+    //   https://m3.material.io/theme-builder#/custom
+    colorScheme: ColorScheme.fromSeed(
+      seedColor: kZulipBrandColor,
+    ),
+    scaffoldBackgroundColor: const Color(0xfff6f6f6),
+    tooltipTheme: const TooltipThemeData(preferBelow: false),
+  );
+}
+
+/// The Zulip "brand color", a purplish blue.
+///
+/// This is chosen as the sRGB midpoint of the Zulip logo's gradient.
+// As computed by Anders: https://github.com/zulip/zulip-mobile/pull/4467
+const kZulipBrandColor = Color.fromRGBO(0x64, 0x92, 0xfe, 1);
+
+/// A [ThemeData] that depends on [outerThemeData] and copies it with additions.
+///
+/// [context] must be below a [Theme] that provides [outerThemeData],
+/// so that this can use [Theme.of] to synthesize a [ThemeData.textTheme]
+/// based on [outerThemeData]'s [ThemeData.typography].
+// TODO(upstream) nicer if this could return a [ThemeData] with just the added
+//   fields, instead of having to call `copyWith` on `Theme.of(context)`.
+//   Then callers could call a hypothetical ThemeData.merge method with it.
+//   But such a method doesn't exist yet:
+//     https://github.com/flutter/flutter/issues/43823
+ThemeData innerThemeData(BuildContext context) {
+  final theme = Theme.of(context);
+  assert(theme.textTheme.labelLarge!.debugLabel!.contains('zulipTypography'));
+  return theme.copyWith(
+    elevatedButtonTheme:  ElevatedButtonThemeData( style: _commonButtonStyle(context)),
+    filledButtonTheme:    FilledButtonThemeData(   style: _commonButtonStyle(context)),
+    iconButtonTheme:      IconButtonThemeData(     style: _commonButtonStyle(context)),
```

Comment:
Seems cleanest to factor this common style out as a local — that way we only compute it once.